Optimization: avoid the pop/putself combo when possible

casperisfine opened the pull request:
NB: all the tests that assert the generated code fail. They're easy to fix but I prefer to gauge interest first before I take the time to do it.

Erubi template generate code that keep putting the buffer back on the stack even though it's already there.

e.g.

```ruby
'<li>'.freeze; _buf << ( message ).to_s; _buf << '</li>
 <li>'.freeze; _buf << ( message ).to_s; _buf << '</li>
 <li>'.freeze; _buf << ( message ).to_s; _buf << '</li>'
```

could be:

```ruby
'<li>'.freeze << ( message ).to_s << '</li>
 <li>'.freeze << ( message ).to_s << '</li>
 <li>'.freeze << ( message ).to_s << '</li>'
```

Every `; _buf` sequence add two VM instruction:

```
0003 pop
0004 putself
```

Both can be avoided when the next instruction is to append to the buffer.

### Benchmark

```
Comparison:
        render patch:   855870.9 i/s
       render 1.10.0:   745255.2 i/s - 1.15x  (± 0.00) slower
```

```ruby
require 'erubi'
require 'benchmark/ips'
erb = Erubi::Engine.new(<<~ERB)
<html>
  <body>
    <ul>
      <li><%= message %></li>
      <li><%= message %></li>
      <li><%= message %></li>
      <% 2.times do |i| %>
        <li><%= i %></li>
      <% end %>
      <li><%= message %></li>
      <li><%= message %></li>
    </ul>
  </body>
</html>
ERB
src = "# frozen_string_literal: true\ndef render(message)\n#{erb.src}\nend"

puts "=" * 40
puts src
puts "=" * 40
eval(src)

Benchmark.ips do |x|
  x.report("render #{Erubi::VERSION}") do
    render("Hello")
  end
  x.save!("/tmp/erubi-bench-data")
  x.compare!
end
```

jeremyevans commented:
Thanks for working on this.  I agree that this optimization makes sense. If you could update the tests to support this, that would be good.  Make sure to run the coverage tests to ensure all branches are covered, and add covering tests if necessary.

casperisfine commented:
Will do!

casperisfine commented:
I updated the test suite:

<img width="824" alt="Capture d’écran 2022-07-26 à 18 12 04" src="https://user-images.githubusercontent.com/19192189/181057196-b2269d2d-63d1-443b-a40a-01fe2642ad47.png">

jeremyevans commented:
Unfortunately, I thought of a case that will break with this, and tested this and found it does break. This breaks cases where the buffer is swapped during template execution (not possible with the default local variable, but possible when an instance variable is used). I'm not sure how common that is, but it at least breaks Roda's chunked plugin.  So this behavior will have to be opt-in and not by default.  Can you pick an option name and make this opt-in?  Maybe `:chain_buffer_calls`?  That is kind of long; I'm open to better ideas.

There are two missed branches after your pull request, directly related to conditionals you added.  Can you please add covering tests for them? Erubi requires 100% line/branch coverage (just like all libraries I maintain).

Can you also remove the WallError debugging code you added to the tests?

paracycle commented:
> Unfortunately, I thought of a case that will break with this, and tested this and found it does break. This breaks cases where the buffer is swapped during template execution (not possible with the default local variable, but possible when an instance variable is used).

Does that mean that the existing test suite does not have tests for those cases? If so, should those be in the test suite so that this PR can make sure that those examples pass without the new option and fail with the new option?

jeremyevans commented:
> > Unfortunately, I thought of a case that will break with this, and tested this and found it does break. This breaks cases where the buffer is swapped during template execution (not possible with the default local variable, but possible when an instance variable is used).
> 
> Does that mean that the existing test suite does not have tests for those cases? If so, should those be in the test suite so that this PR can make sure that those examples pass without the new option and fail with the new option?

Correct, there are no current tests for the behavior.  However, since Erubis and ERB both appear to not chain buffer calls, and there is external code relying on buffer calls not being chained, Erubi should continue to not chain by default.

I agree that it would be good to add tests where the buffer variable is modified (which relies on the non-chaining implementation).  However, I don't think we should add a test that modifies the buffer variable and enables chaining, and specifically checks that case fails.

casperisfine commented:
Ok, I updated the PR:

  - I named the option `immutable_bufvar`. I'm not 100% happy with it, but I couldn't find anything better. It's trying to convery that `bufvar` is never re-assigned.
  - I added a specific test for it, now coverage report 100% of branches covered. Let me now if you'd like to test more corner cases.
  - This had the side effect of changing the generated code even when the option is not enabled, it cause a bunch of unnecessary semi-colon, which I assume is fine. If not let me know and I can try to write some smarter code to keep the old code shape, but otherwise I think dumb and easy to grasp generator code trump cleaner generated code.

jeremyevans commented:
`immutable_bufvar` sounds to me like the bufvar is immutable, which is definitely not the case as the content appends to it.  But I can easily change the option name.  Maybe `chainable_bufvar` or `chain_appends`? 

Thanks for working on the branch coverage.

The style of the generated code isn't too much of a concern.  Extra semicolons are no-ops in terms of VM instructions, so having them shouldn't cause any problems.

byroot commented:
chain_appends sound good to me. I can update the PR tomorrow or feel free to do it post merge. Let me know of there’s any other concern

jeremyevans commented:
I've been working with this code for the last couple hours, and to reduce the potential for breaking subclasses, I've modified things so that the code generated by each method is the same as before if the `:chain_appends` option is not used.  The only change in the tests is the addition of a test for the new option, and a minor tweak to the `freeze` handling.

byroot commented:
👏
